add HCAL laser monitoring charge

jkunkle opened the pull request:
Added a float and a getter function to HcalNoiseSummary to store/retrieve the laser monitoring energy

Added code in HcalNoiseInfoProducer to calculate the charge and add it to the noise summary

The inputs to the calculation are provided in hcalnoiseinfoproducer_cfi.py : 

- A flag to turn the calculation on or off
- three vectors to specify which channels to use in time order
  - detType
  - iEta
  - iPhi
- A minimum and maximum for the integration limits.  By default the integral is performed over the entire range

Commit:
add laser monitor charge to HcalNoiseSummary which is calculated in HcalNoiseInfoProducer

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/1801

cmsbuild commented:
A new Pull Request was created by  @jkunkle  for master.

It involves the following packages:

DataFormats/METReco
RecoMET/METProducers


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @ahinzmann, @mmarionncern, @rappoccio, @rovere, @jdamgov, @jdolen, @nhanvtran, @gkasieczka, @schoef, @mariadalfonso, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 76792814983aa73bb11a456bfa4127665b39d656


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/2b5a412244b636d4dcc20dd832de2195276df259
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24168/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24168/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24168/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
7.3 step2
<pre>runTheMatrix-results/7.3_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17/step2_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17.log

</pre>136.788 step2
<pre>runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step2_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

</pre>10042.0 step2
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>10024.0 step2
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>10224.0 step2
<pre>runTheMatrix-results/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU.log

</pre>11624.0 step2
<pre>runTheMatrix-results/11624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019.log

</pre>10824.0 step2
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v31.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:31:46 2017-date Fri Nov  3 22:21:38 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v31,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v31 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v31_MC.root --fileout file:RelVal_Raw_2e34v31_MC_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:31:46 2017-date Fri Nov  3 22:21:38 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:28:15 2017-date Fri Nov  3 22:21:51 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:28:15 2017-date Fri Nov  3 22:21:51 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:32:33 2017-date Fri Nov  3 22:21:55 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_GRun_MC.root --fileout file:RelVal_Raw_GRun_MC_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:32:33 2017-date Fri Nov  3 22:21:55 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v40.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:40:01 2017-date Fri Nov  3 22:28:25 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v40,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v40 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v40_MC.root --fileout file:RelVal_Raw_2e34v40_MC_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:40:01 2017-date Fri Nov  3 22:28:25 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v31.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:37:44 2017-date Fri Nov  3 22:29:36 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v31,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_2e34v31 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v31_DATA.root --fileout file:RelVal_Raw_2e34v31_DATA_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:37:44 2017-date Fri Nov  3 22:29:36 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v22.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:44:22 2017-date Fri Nov  3 22:31:56 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v22,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v22 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v22_MC.root --fileout file:RelVal_Raw_2e34v22_MC_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:44:22 2017-date Fri Nov  3 22:31:56 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v22.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:39:41 2017-date Fri Nov  3 22:32:38 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v22,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_2e34v22 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v22_DATA.root --fileout file:RelVal_Raw_2e34v22_DATA_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:39:41 2017-date Fri Nov  3 22:32:38 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02496/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_0_X_2017-11-03-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v40.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Fri Nov  3 22:51:33 2017-date Fri Nov  3 22:42:14 2017 s - exit: 18688
cmsDriver.py RelVal -s HLT:2e34v40,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_2e34v40 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v40_DATA.root --fileout file:RelVal_Raw_2e34v40_DATA_HLT_RECO.root : FAILED - time: date Fri Nov  3 22:51:33 2017-date Fri Nov  3 22:42:14 2017 s - exit: 18688



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/2b5a412244b636d4dcc20dd832de2195276df259
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24168/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24168/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

perrotta commented on DataFormats/METReco/src/classes_def.xml:
```diff
@@ -133,7 +133,7 @@
    <version ClassVersion="13" checksum="382065414"/>
    <version ClassVersion="14" checksum="651605939"/>
-   <version ClassVersion="15" checksum="2415916091"/>
+   <version ClassVersion="15" checksum="1886001321"/>
```
Please, put back the previous class version, and move to a new one ("16") because of the added class member

perrotta commented on RecoMET/METProducers/interface/HcalNoiseInfoProducer.h:
```diff
@@ -110,6 +111,7 @@ namespace reco {
     edm::EDGetTokenT<reco::PFJetCollection> jet_token_;
 
     double TotalCalibCharge;    // placeholder to calculate total charge in calibration channels
+    double TotalLasmonCharge;    // placeholder to calculate total charge in laser monitor channels
```
Class members should start with lowercase letters, uppercase initials are reserved to class names
(Please to the same also for TotalCalibCharge, and other possible occurrencies in this class)

perrotta commented on RecoMET/METProducers/interface/HcalNoiseInfoProducer.h:
```diff
@@ -130,8 +132,16 @@ namespace reco {
     uint32_t HcalAcceptSeverityLevel_;
     std::vector<int> HcalRecHitFlagsToBeExcluded_;
+
+    std::vector<int> LaserMonDetTypeList_;
```
Class members should start with lowercase letters, uppercase initials are reserved to class names

abdoulline commented:
email with a (fixing) recipe sent to internal HCAL thread...

perrotta commented on RecoMET/METProducers/interface/HcalNoiseInfoProducer.h:
```diff
@@ -130,8 +132,16 @@ namespace reco {
+    std::vector<int> LaserMonIPhiList_;
+    std::vector<int> LaserMonIEtaList_;
+
+    int LaserMonitorTSStart_;
```
Class members should start with lowercase letters, uppercase initials are reserved to class names

perrotta commented on RecoMET/METProducers/python/hcalnoiseinfoproducer_cfi.py:
```diff
@@ -82,7 +82,14 @@
-                                    0.0, 0.0, 1.0, -0.2)   # equivalent to "energy fraction > 0.2"
+                                    0.0, 0.0, 1.0, -0.2),   # equivalent to "energy fraction > 0.2"
+
+    LaserMonDetTypeList = cms.vint32(3,3,3,3,3,3,3,3),
```
Lower case initials also for config parameters

perrotta commented on RecoMET/METProducers/python/hcalnoiseinfoproducer_cfi.py:
```diff
@@ -82,7 +82,14 @@
+    LaserMonIPhiList = cms.vint32(23,17,11,5,29,35,41,47),
+    LaserMonIEtaList = cms.vint32(0,0,0,0,0,0,0,0),
+
+    #LaserMonitorTSStart = cms.int32( 4 ),
```
Are these commented out parameters needed: if so, please add a comment explaining why you keep them commented out, otherwise just remove

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -112,6 +113,49 @@ HcalNoiseInfoProducer::HcalNoiseInfoProducer(const edm::ParameterSet& iConfig) :
   }
 
+  // get the fiber configuration vectors
+  std::vector<int> TmpLaserMonDetTypeList = iConfig.getParameter<std::vector<int> >("LaserMonDetTypeList");
```
Local variables should start with lowercase letters, as well as the config parameters

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -458,6 +529,41 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+
+                    // now get the digis
+                    int ts_size = int(digi->size());
+	            for(int i = 0; i < ts_size; i++) {
```
Correct indentation

perrotta reviewed (commented):
While applying the fix suggested in https://github.com/cms-sw/cmssw/pull/21152#issuecomment-341876046, please take also into account the following (preliminary) review comments

slava77 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -112,6 +113,49 @@ HcalNoiseInfoProducer::HcalNoiseInfoProducer(const edm::ParameterSet& iConfig) :
+  }
+
+  // get the integration region with defaults
+  if( iConfig.existsAs<int>("LaserMonitorTSStart" ) ) {
```
use of existsAs should be justified, because it is an exception.
Default values for parameters should be defined with fillDescriptions.
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

perrotta commented:
@jkunkle : please update about the status of the implementation of the bug fix mentioned in https://github.com/cms-sw/cmssw/pull/21152#issuecomment-341876046, and the implementation of the preliminary review comments provided in https://github.com/cms-sw/cmssw/pull/21152#pullrequestreview-74248374 and https://github.com/cms-sw/cmssw/pull/21152#pullrequestreview-74256490

Could you please also post here a recipe for testing the effects of this PR, once fixed?

perrotta commented:
@junkle : is there any update?

Will this Pull Request get discussed by @mariadalfonso (or some other HCAL people) this week at the reco meeting? If not, maybe you could come yourself and discuss it: we can allow one slot for it, just let us know (and please contact me, perrotta@vxcern.cern.ch, to arrange for it)

Junkle commented:
@perrotta Hi, someone appears to have accessed my account without my authorization.  Please disregard any previous actions from it.

Commit:
update to include fillDescriptions and the METFilter code

Commit:
remove printout

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/1980

cmsbuild commented:
Pull request #21152 was updated. @ghellwig, @lveldere, @ianna, @kpedro88, @fabozzi, @rekovic, @vanbesien, @govoni, @perrotta, @civanch, @monttj, @cmsbuild, @GurpreetSinghChahal, @davidlange6, @smuzaffar, @Dr15Jones, @efeyazgan, @mdhildreth, @jfernan2, @cerminar, @slava77, @ggovi, @mulhearn, @ssekmen, @prebello, @vazzolini, @kmaeshima, @arunhep, @perrozzi, @thuer, @dmitrijus, @franzoni, @lpernie can you please check and sign again.

jkunkle commented:
My original intent was to make as few modifications to the code, but its clear that this code is outdated in functionality and style.  Therefore I'll make a few more changes to accommodate the suggestions.  I have pushed the updated code to the branch.  The updates include,

- add a fillDescriptions function to RecoMET/METProducers/src/HcalNoiseInfoProducer.cc and RecoMET/METProducers/include/HcalNoiseInfoProducer.h which sets the defaults based on the values set in RecoMET/METProducers/python/hcalnoiseinfoproducer_cfi.py
- change the variable member names to match the suggested style, where appropriate
- include also the changes to RecoMET/METFilters/RecoMET/METFilters/plugins/HcalLaserEventFilter.cc to set the hcal laser MET filter based on the stored value
- other minor changes suggested above

I've developed some simple testing code.  I've put a RAW data file which contains the events of interest on my eos space.  You can run on the RAW code, process that into a miniAOD file, and then run a simple analyzer.  The analyzer counts the number of events that pass and fail the laser filter and compares them directly to a cut applied on the laser monitor charged as stored in HcalNoiseSummary.  A printout at the end of the last step shows this comparison.  By default I run over 100 events, but this can be changed in the config files.  The steps to setup and run the test are, 

cmsrel CMSSW_10_0_X_2017-11-12-0000
cd CMSSW_10_0_X_2017-11-12-0000/src
cmsenv
git cms-merge-topic 21152
cp -R /afs/cern.ch/user/j/jkunkle/public/HcalLaserCodeTest/* . 
scram b -j8
cmsRun reco_RAW2DIGI_RECO.py
cmsRun aod_PAT.py
cmsRun TestLaserCode/TestLaserCode/test/config_test.py

perrotta commented:
@jkunkle : rebase needed

See http://cms-sw.github.io/tutorial-resolve-conflicts.html

jkunkle commented:
@perrotta  I run git cms-rebase-topic 21152 on a clean checkout and get no modifications to my branch : 
git push -f my-cmssw laserMonitoring10X
Everything up-to-date

Where do you see a merge conflict?  Thanks.

perrotta commented:
From the fact that in this PR there are now 183 commits and 2586 modified files...

abdoulline commented:
Josh, you have an alternative:

option (1) not very elegant, but rather straightforward:
to close this PR and to re-submit from one of the most recent IBs 
after moving manually (cherry-picking) all your updated files to your 
new IB developer area

option (2) elegant, but could be rather tricky (first time) :
try follow Kevin's recipe
http://cms-sw.github.io/tutorial-resolve-conflicts.html



On Mon, 13 Nov 2017, jkunkle wrote:

> 
> @perrotta I run git cms-rebase-topic 21152 on a clean checkout and get no
> modifications to my branch :
> git push -f my-cmssw laserMonitoring10X
> Everything up-to-date
> 
> Where do you see a merge conflict? Thanks.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AEx02nkKoH4soqjfL8oMebm_gxHft_fNks5s2Ki6gaJpZM4QRXb5.gif]
> 
> 
>

abdoulline commented:
...sorry, forgot to add a quote from the aforementioned recipe:

<...>
Very important note: NEVER use git cms-merge-topic to resolve conflicts in 
a pull request!!! This can introduce hundreds of duplicate commits. Release 
and package managers have the right and the responsibility to reject such 
broken pull requests immediately.
<...>


On Mon, 13 Nov 2017, Salavat Abdoulline wrote:

>
>  Josh, you have an alternative:
>
> option (1) not very elegant, but rather straightforward:
> to close this PR and to re-submit from one of the most recent IBs after 
> moving manually (cherry-picking) all your updated files to your new IB 
> developer area
>
> option (2) elegant, but could be rather tricky (first time) :
> try follow Kevin's recipe
> http://cms-sw.github.io/tutorial-resolve-conflicts.html
>
>
>
> On Mon, 13 Nov 2017, jkunkle wrote:
>
>>
>>  @perrotta I run git cms-rebase-topic 21152 on a clean checkout and get no
>>  modifications to my branch :
>>  git push -f my-cmssw laserMonitoring10X
>>  Everything up-to-date
>>
>>  Where do you see a merge conflict? Thanks.
>>
>>  —
>>  You are receiving this because you are subscribed to this thread.
>>  Reply to this email directly, view it on GitHub, or mute the
>>  thread.[AEx02nkKoH4soqjfL8oMebm_gxHft_fNks5s2Ki6gaJpZM4QRXb5.gif]
>>
>> 
>> 
>

cmsbuild commented:
The code-checks are being triggered in jenkins.

jkunkle commented:
Ok, I was following the recipe, but I must have messed it up.  It seems to have worked as expected this time

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/1982

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8141e8cbd99d01c06da6381e1154499c5c393305
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24407/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/c738e9e717b9f5ee025a84a1e1c8c20b9084dddb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24407/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24407/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24407/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2832699
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2832520
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.05999999997 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

abdoulline commented:
@jkunkle you did it, thank you, Josh.

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -452,15 +696,50 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
         for(int i = 0; i < (int)digi->size(); i++)
-           TotalCalibCharge = TotalCalibCharge + tool[i];
+           totalCalibCharge = totalCalibCharge + tool[i];
         */
```
There is a large chunk of commented out code here above: couldn't it be just removed?
Otherwise, please add some further comment which explains why do you want to keep this commented out code here

jkunkle replied:
The current comment, 

   // Original code computes total calib charge over all digis.  While I think it would be more useful to skip
        // zs mark-and-pass channels, I keep this computation as is.  Individual HBHE and HF variables do skip
        // the m-p channels.  -- Jeff Temple, 6 December 2012

indicates that its quite old.   I can remove the commented code if desired

perrotta replied:
Yes, please. Unless there is a reason to maintain it there (which I don't think)

Commit:
fix checksum, tabulation, and remove commented code

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/1988

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f7507af4ff4fdbf4427b849fb5087767e2016313
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24422/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/45cd82f270d6a7ba7cec0c29d789fba271283d74
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24422/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24422/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24422/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2716578
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2716401
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.03000000012 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +753,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+       matching_idx[nFibers - 1] = 10;
+
+       // now loop over the time slices of each fiber and make the sum
+       int icombts = -1;
```
I think that laserMonitorTSStart_ and laserMonitorTSEnd_ are meant to bound the time slices inside each fiber: if so, this initialization should be moved inside the first loop below.

jkunkle replied:
The aim of TSStart and TSEnd are to provide boundaries (if needed) for the charge integration after the data from the various fibers have been stitched together.   Therefore icombts should count from 0 to the total number of TS (which could be ~60)  and the limits evaluated based on that count

perrotta replied:
Thank you for having checked and told.
The logic is not straightforward to follow, thugh... But ok, if you confirn that it is correct as such this is fine.

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -112,6 +113,40 @@ HcalNoiseInfoProducer::HcalNoiseInfoProducer(const edm::ParameterSet& iConfig) :
+  std::vector<int> tmpLaserMonIEtaList = iConfig.getParameter<std::vector<int> >("laserMonIEtaList");
+
+  // the transfer of data from python to c seems to have issues
+  // this can be fixed by explicitly filling the vector
```
???

jkunkle replied:
There seems to be some memory issue here that I wasn't able to resolve otherwise.  If I simply assign the member variables directly with getParameter (and eliminate the code that push_backs the vectors) and then print the values, it looks OK.  However, in the produce() function I find that the vectors are empty (ie ( laserMonIEtaList_.size() == 0 ) evaluates to True and the same for the other two vectors).  Have you seen this before?

Dr15Jones replied:
What exactly was the code you were trying to use previously? That would help diagnose what was happening.

jkunkle replied:
I've made a simple patch file that you can apply as below, recompile, and do cmsRun reco_RAW2DIGI_RECO.py (from my example setup above) to reproduce my printout.

patch RecoMET/METProducers/src/HcalNoiseInfoProducer.cc  -i /afs/cern.ch/user/j/jkunkle/public/HcalLaserCodeTest/printtest.patch  -o HcalNoiseInfoProducer.cc
mv HcalNoiseInfoProducer.cc RecoMET/METProducers/src/
scram b
cmsRun reco_RAW2DIGI_RECO.py 

You should see in the stdout

---

Vector sizes in constructor : 
laserMonIPhiList_.size() = 8
laserMonIEtaList_.size() = 8
laserMonDetTypeList_.size() = 8
---
Vector sizes in produce : 
laserMonIPhiList_.size() = 0
laserMonIEtaList_.size() = 0
laserMonDetTypeList_.size() = 0

Dr15Jones replied:
The problem is here in the constructor:
```C++
  std::vector<int> laserMonDetTypeList_= iConfig.getParameter<std::vector<int> >("laserMonDetTypeList");
  std::vector<int> laserMonIPhiList_   = iConfig.getParameter<std::vector<int> >("laserMonIPhiList");
  std::vector<int> laserMonIEtaList_   = iConfig.getParameter<std::vector<int> >("laserMonIEtaList");
```

you redefined the variable names to be function local variables instead of member variables. The fix is to remove the type declaration:

```C++
  laserMonDetTypeList_= iConfig.getParameter<std::vector<int> >("laserMonDetTypeList");
  laserMonIPhiList_   = iConfig.getParameter<std::vector<int> >("laserMonIPhiList");
  laserMonIEtaList_   = iConfig.getParameter<std::vector<int> >("laserMonIEtaList");
```

jkunkle replied:
ah my mistake, I should have seen that.  Thanks

perrotta reviewed (commented):
Since you define and validate now all config parameters via fillDescriptions, please remove all unneeded existsAs checks in HcalNoiseInfoProducer.cc

perrotta commented:
Thank you @jkunkle for having provided the test configs: I could use them and verify as such the functioning of the filter.

Still, after having reviewed better the code, I think there is a bug at line 861 of HcalNoiseInfoProducer.cc, which does not affect the test run because of the values of the parameter laserMonitorTSStart and laserMonitorTSEnd given now in the configuration. Please check and either fix or tell me that Im wrong.

Please also take care or answer the other two small points raised in the review https://github.com/cms-sw/cmssw/pull/21152#pullrequestreview-77068928

After that the PR could be ready to be signed off by reco.

Commit:
remove remaining existsAs, fix hack in filling vectors

cmsbuild commented:
The code-checks are being triggered in jenkins.

jkunkle commented:
I've committed the two changes suggested
1) remove remaining existsAs 
2) fix the hack in filling the vectors

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2034

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 25362e63444e0a2844e79d379811de42e5dfad58
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24494/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/c8c0943eda6b18961c7c9b682b7770ddeba56b92
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24494/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24494/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24494/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2833444
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2833265
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.44999999998 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

perrotta commented:
+1
- Functioning of the filter verified with the provided test config
- Some further detail on the algorithm available in https://indico.cern.ch/event/677823/contributions/2775506/attachments/1559532/2454479/monitoring_v0.pdf : this answers my doubts in https://github.com/cms-sw/cmssw/pull/21152#discussion_r151373909

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -125,6 +124,19 @@ HcalNoiseInfoProducer::HcalNoiseInfoProducer(const edm::ParameterSet& iConfig) :
      adc2fC[i] = adc2fCTemp[i];
 
+  // adc -> fC for qie8 with PMT input, for laser monitor
+  const float adc2fCTempHF[128]={-3,-0.4,2.2,4.8,7.4,10,12.6,15.2,17.8,20.4,23,25.6,28.2,30.8,33.4,
```
hi @jkunkle - i don't think that this intermediate variable is needed. you should be able to just do

adc2fCHF = {..}

otherwise, please add some documentation of the source of these numbers (that should be in a conditions db..)

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -428,39 +623,61 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
   {
+
+     // store the data from the lasermon fibers
+     std::map<int, std::vector<int> > lasmon_adcs;
```
it looks like these can be vectors instead of maps
std::vector<std::vector<int> > lasmon_capids(laserMonDetTypeList_.size(), std::vector<int>());

jkunkle replied:
Yes, I agree.  The map was intended to be more flexible, but it ends up that we just loop in order

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -428,39 +623,61 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+
+                    // now get the digis
+                    int ts_size = int(digi->size());
+                    for(int i = 0; i < ts_size; i++) {
```
generic question - all the array indices are better kept as unsigned ints. Why make them ints?

jkunkle replied:
I switched them to unsigned

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +717,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+       // them so we dont run into problems later
+       for( unsigned idx = 0; idx < laserMonDetTypeList_.size(); ++idx ) {
+           if( lasmon_adcs[idx].empty() ) {
+               for( int i = 0; i < 10; ++i ) {
```
this can be written as lasmon_adcs[idx]=std::vector<int>(10,-1)

jkunkle replied:
yes, done

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +717,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+       // data from the next fiber matches in order to stitch them together.
+       // When there is an overlap, the data from the end of the
+       // earlier fiber is removed.  There is no removal of the last fiber
+       std::map<int, int> matching_idx; 
```
it looks like this can be a vector of ints

jkunkle replied:
As in the other maps, it seems that indeed the map is not necessary.  I've changed it

Commit:
add code changes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2096

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented:
plese test

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,130 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+           int nmatch_cap = 0; // count the number of TS where capID matched
+           int nmatch_adc = 0; // count the number of TS where ADC matched
+
+           for( int cidx = start_ts, nidx = 0; cidx <= last_ts; cidx++, nidx++ ) { 
```
At this point, also cidx and nidx could be unsigned

jkunkle replied:
nidx can be unsigned, but since start_ts needs to be allowed to go negative cidx should be an int.  I changed nidx to be unsigned

perrotta replied:
start_ts is kept >= 0 by the while statement at L743

jkunkle replied:
the while statement only executes if start_ts >=0.  start_ts is decremented within the while loop if there is no match found.  in the case when the loop starts with start_ts == 0, it can therefore go to -1.  This is a special case which indicates that no match was found between two channels and is handled after the while loop

perrotta replied:
When cidx is assigned start_ts, such a start_ts is not yet decremented.
Therefore cidx can't be assigned something below 0 by construction

jkunkle replied:
right, the indices themselves do not go below 0, but start_ts and last_ts should be signed to be consistent with their implementation.  therefore I can do this,

  for( unsigned cidx = (unsigned)start_ts, nidx=0; cidx <= (unsigned)last_ts; cidx++, nidx++ ) {

is this satisfactory?

perrotta replied:
Not nice, indeed... On the other hand, also having a few indexes signed and a few unsigned in not so nice. About the performance I don't think it is an issue anyhow.

@davidlange6 : what do you suggest?

If there are no other suggestions, I will start the test with what will be in this PR in a couple of hours (as I will be mostly online in the meanwhile...)

davidlange6 replied:
i had missed that - one could use a different "default" value - eg, std::numeric_limits<unsigned int>::max()

davidlange6 replied:
sorry - ignore that - I think I didn't follow completely the logic yet

jkunkle replied:
I went ahead and updated it to what I suggested.  Let me know if you have other suggestions off of the most recent pull

davidlange6 replied:
thinking outloud - doing

unsigned int tmp=0;
while ( tmp < last_ts+1 ) {
   unsigned int start_ts = last_ts-i;
 ..
}
if ( tmp > last_ts)...

would remove all the mixups between int and unsigned int. I didn't try to implement it so maybe its more effort than benefit. 

That said, this logic has a lot of places where "off-by-one" errors can come in. Its good to review it to see if there is a simpler approach.

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,130 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+           for( int cidx = start_ts, nidx = 0; cidx <= last_ts; cidx++, nidx++ ) { 
+             ncheck++;
+             // if we get an invald value, move on
+             if( lasmon_capids[fidx][cidx] == -1 ) continue;
```
You can just break here: if( lasmon_capids[fidx][cidx] == -1 )  there is no possibility any more that nmatch_cap and nmatch_adc can become equal to ncheck

jkunkle replied:
yes, done

Commit:
switch one int to unsigned, change continue to break

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2103

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

jkunkle commented:
Apologies for adding extra noise, but I've found a mistake in the code which I will update in the next push of the branch.  The issue is that I wasn't using the full functionality of the new entries in the emap by using the hcalSubdet() accessor of the HcalCalibDetId.  Instead the cboxChannel() accessor should be used.  This implies these changes

- change the accessor in line 636 of HcalNoiseInfoProducer.cc to cboxChannel()
- rename detid to cboxch and laserMonDetIdList to laserMonCBoxList for both the member variables and the python names
- change the defaults from 3 to 6 in the python and fillDescriptions()

The logic of the code does not change and the functionality is the same.  Using cboxChannel is more robust against emap changes though since we explicitly plan to use these cboxChannel vales of 6 for these channels

Commit:
switch to using cboxChannel instead of hcalSubdet to identify laser monitor channels.  The default value is updated to 6, and names are changed to avoid confusion.  The for loop has been changed to unsigned for now

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2123

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

Commit:
simplify the loop to determine matching time slices

jkunkle commented:
I've pushed a proposed simplification to the code in question.  switched from the while loop to a for loop over the number of TS to check.  I then just loop from the 0th TS to the number to check.  The tradeoff is in determining the index of the earlier fiber (in time) to check against the later fiber.  I've renamed the indices to make it clear which is the later fiber (lidx) or earlier fiber (eidx).  I've run the usual test and confirmed that the calculation is correct.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2144

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -730,40 +730,42 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+           for( unsigned lidx = 0; lidx < ncheck; lidx++) {
+             // we are looping over the TS of the later fiber in time
+             // the TS of the earlier fiber starts from the end
+             unsigned eidx = last_ts-ncheck+lidx+1;
```
```
            unsigned eidx = last_ts-ncheck+lidx;
```
In principle this is risky, but ncheck is bound to last_ts at L744, therefore it works

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+       // we assume that the list of fibers was given in time order
+       // (if this was not the case, then we just end up using 
+       // all data from all fibers )
+       for( unsigned fidx = 0; fidx < (nFibers - 1); ++fidx ) {
```
This fails if nFibers=0, which is not allowed by construction, but it is anyhow worth putting a protection:
```
       for( unsigned fidx = 0; nFibers>0 && fidx < (nFibers - 1); ++fidx ) {
```

jkunkle replied:
OK

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+         // on each iteration check if all capId and ADCs from this
+         // TS forward match the beginning entries of 
+         // the next fiber
+         unsigned last_ts = lasmon_capids[fidx].size()-1;  // last TS
```
This fails if size()=0
Better define here
```
        unsigned last_ts = lasmon_capids[fidx].size();  // last TS
```
and use it coherently in the following

jkunkle replied:
This should work with the adjustments that you suggest.  To maintain clarity I would rename last_ts to nts

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+         // TS forward match the beginning entries of 
+         // the next fiber
+         unsigned last_ts = lasmon_capids[fidx].size()-1;  // last TS
+         int start_ts = last_ts; // start_ts will be decrimented on each loop
```
```
         int start_ts = last_ts-1; // start_ts will be decrimented on each loop
```

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+         // store the latest capID that has a match
+         int latest_cap_match = -1;
+         // reverse loop over TSs
+         for( unsigned ncheck = 1; ncheck <= lasmon_capids[fidx].size() ; ncheck++ ) {
```
```
        for( unsigned ncheck = 1; ncheck <= last_ts ; ncheck++ ) {
```

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+             // N-4 spot (and the ADCs will not)
+             // if this is not the case, then we just take
+             // the value of latest match
+             if( latest_cap_match == int(last_ts - 3) ) {
```
```
             if( latest_cap_match == int(last_ts - 4) ) {
```

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+           // if we didn't find any match, use the capID only to compare
+           if( latest_cap_match < 0 ) {
+             //this shouldn't happen, in this case use all the data from the fiber
+             start_ts = lasmon_capids[fidx].size();
```
```
               start_ts = last_ts;
```

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+             // if this is not the case, then we just take
+             // the value of latest match
+             if( latest_cap_match == int(last_ts - 3) ) {
+               start_ts = lasmon_capids[fidx].size();
```
```
               start_ts = last_ts;
```

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,134 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+         int latest_cap_match = -1;
+         // reverse loop over TSs
+         for( unsigned ncheck = 1; ncheck <= lasmon_capids[fidx].size() ; ncheck++ ) {
+           unsigned nmatch_cap = 0; // count the number of TS where capID matched
```
Probably, the whole block between L745 and L767 could be simplified as follow (without changing the logic):
```
           // loop over the channel TS, this is for the later fiber in time
           bool cap_match = true;
           bool adc_match = true;
           for( unsigned lidx = 0; lidx < ncheck; lidx++) {
             // we are looping over the TS of the later fiber in time
             // the TS of the earlier fiber starts from the end
             unsigned eidx = last_ts-ncheck+lidx;
             // if we get an invald value, this fiber has no data
             // the check and match will fail, so the start_ts will 
             // be decrimented
             if( lasmon_capids[fidx][eidx] == -1 || lasmon_capids[fidx+1][lidx] == -1 ) {
               cap_match = false;
               adc_match = false;
	       break;
	     }
             if( lasmon_capids[fidx][eidx] == lasmon_capids[fidx+1][lidx] ) {
	       cap_match = false;
             }
             // check the data values as well
             if( lasmon_adcs[fidx][eidx] == lasmon_adcs[fidx+1][lidx] ) {
	       adc_match = false;
             }
           }
```

jkunkle replied:
I think it is possible.  Whenever a match fails the bool is set to false.  I think in the code above you mean to have != instead of == 

             if( lasmon_capids[fidx][eidx] != lasmon_capids[fidx+1][lidx] ) {
	       cap_match = false;
             }
             // check the data values as well
             if( lasmon_adcs[fidx][eidx] != lasmon_adcs[fidx+1][lidx] ) {
	       adc_match = false;
             }

perrotta replied:
Yes, indeed...
Thank you for noticing

perrotta reviewed (commented):
There are a few places in your latest commit in which you make differences between unsigned quantities, which is prone to errors. Please find here a few suggestions about how to get rid of them, without changing the logic of your implementation

Commit:
move last_ts to nts, instead of counting matches, just look for a mismatch, improve the comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

jkunkle commented:
the code has been updated to reflect the latest comments.  I also improved some of the comments in the code.  The test shows the expected behavior

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2155

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -500,9 +707,141 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+             // if this is not the case, then we just take
+             // the value of latest match
+             if( latest_cap_match == int(nts - 4) ) {
+               start_ts = lasmon_capids[fidx].size();
```
```
start_ts = nts;
```
(as in L795 above...)

jkunkle replied:
ah, missed that one

Commit:
807 : start_ts = nts;

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2157

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 52993b99e55e131190c2a4b40e9bcbf0f8f9a68a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24630/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/5780cd48e28d35f8052219e1dd10cb8f9fae4c43
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24630/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24630/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24630/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2833444
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2833265
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.41999999995 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

perrotta commented:
+1
- Already signed in https://github.com/cms-sw/cmssw/pull/21152#issuecomment-345313427
- Further updates, originally requested by @davidlange6, improved the code further
- Verified with the test script provided that the code behaves as expected even after the latest updates

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -428,39 +621,53 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+              if( std::find( laserMonIEtaList_.begin(), laserMonIEtaList_.end(),
+                             ieta ) != laserMonIEtaList_.end() ) {
+                // we have a lasmon channel, find the index in the list of inputs
+                for( unsigned idx = 0; idx < laserMonCBoxList_.size(); ++idx ) {
```
hi @jkunkle - sorry, I thought I had commented on this before, but it seems not... these finds do not seem to be needed given the loop that follows here. The check in the loop takes advantage of the knowledge that the index is always the same.

jkunkle replied:
Hi @davidlange6 .  The point here is to only run the loop for the 4-8 channels of interest rather than for every calibration channel.  These std::finds could easily be removed, but the loop will run for every channel.  Alternatively one could just keep the cBoxList check, which would be sufficient to filter out these channels.  What do you think?

abdoulline commented:
@jkunkle
Josh, once you've put complete fillDescriptions() in place, you don't need to (even shouldn't)  keep full-fledged (redundant) content in hcalnoiseinfoproducer_cfi.py

As all the use-cases (consumers of "hcalnoise" producer)
http://cmslxr.fnal.gov/search?_filestring=&_string=hcalnoiseinfoproducer_cfi&_casesensitive=1
use
import RecoMET.METProducers.hcalnoiseinfoproducer_cfi
(assuming "hcalnoise" producer defined in it)
to preserve them (consumers) intact, you need to keep 1(one) "import" line in
hcalnoiseinfoproducer_cfi.py
(in addition to usual "title":  import FWCore.ParameterSet.Config as cms) :

from RecoMET.METProducers.hcalnoise_cfi import *  # from fillDescriptions

(it works)

perrotta commented:
Good point @abdoulline , I overlooked it.

Due to that, I have removed the "+1" waiting for the redundant content in hcalnoiseinfoproducer_cfi.py getting removed

To do so, in the hcalnoiseinfoproducer_cfi.py:
'''
from RecoMET.METProducers.hcalnoise_cfi import * # from fillDescriptions
hcalnoiseinfoproducer = RecoMET.METProducers.hcalnoise_cfi.hcalnoise.clone()
'''
(followed by possible further adaptations if the hcalnoise configuration has any different parameter than the wanted hcalnoiseinfoproducer one).

While doing so, please profit and rewrite in a more readable way the fillDescription method of HcalNoiseInfoProducer:
- Reorder the parameters by grouping them in a more logical way, for example as it is done now in hcalnoiseinfoproducer_cfi.py
- Write in a single line the short vectors of input parameters, or in a few lines slightly longer ones (for example as it is done now in hcalnoiseinfoproducer_cfi.py), instead of having long lists with just one parameter per line
- Perhaps, you can also profit of the possibility of putting the same comments as e.g. now in hcalnoiseinfoproducer_cfi.py in the auto-generated  hcalnoise_cfi.py, see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Adding_Comments

perrotta commented:
(I have updateted the previous comment of mine)

Commit:
remove all defaults in hcalnoiseinfoproducer_cfi, reorder and add comments in HcalNoiseInfoProducer.cc

jkunkle commented:
I have made the updates suggested above including,

1) remove all defaults in hcalnoiseinfoproducer_cfi
2) reorder entries in fillDescriptions in HcalNoiseInfoProducer.cc
3) add comments in fillDescriptions in HcalNoiseInfoProducer.cc
4) I also implemented my suggestion of only making one std::find check using just the cBoxChannel

perrotta commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21152/2372

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21152 was updated. @perrotta, @cmsbuild, @monttj, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 4d7e1df36f513ea76a1abae0781437d562aeb523
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24796/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/a2fb2580031394d10dfcaa536c13033cb75af162
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24796/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24796/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21152/24796/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834906
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.20999999998 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

perrotta commented:
+1
- Already signed in https://github.com/cms-sw/cmssw/pull/21152#issuecomment-345313427
- Latest additions include fixes following David's comments, and the removal of duplicate explicit cfi (also defined with the fillDescriptor, which has also been improved in readability here)

davidlange6 commented on RecoMET/METProducers/src/HcalNoiseInfoProducer.cc:
```diff
@@ -428,39 +593,45 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
+          int ieta    = digi->id().ieta();
+         
+          // only check channels having the requested cboxch
+          if (std::find( laserMonCBoxList_.begin(), laserMonCBoxList_.end(),
```
just to clarify the previous discussion - "find" is effectively looping over your container - so its not that you are just looking at the 4-8 channels of interest, your looking at all of them (and 4-8 of them, you are looking twice)

jkunkle replied:
Right, there are two cases
For the channels of interest the std::find should return true on the first iteration and then perform the inner loop
For all other channels the std::find makes one loop and returns false

so yes, its clear that the std::find does not help, but it does not hurt much.  I can remove it if you like

davidlange6 commented:
merge

slava77 commented on DataFormats/METReco/interface/HcalNoiseSummary.h:
```diff
@@ -194,6 +195,7 @@ class HcalNoiseSummary
   double rechitEnergy_;
   double rechitEnergy15_;
   double calibCharge_;
+  double lasmonCharge_;
```
this variable is not initialized.
Since I've added the MET filter path monitoring a few days ago, now we find that this leads to random differences when reading AOD produced by a release before 100X.
E.g. https://github.com/cms-sw/cmssw/pull/25320#issuecomment-441199764

@perrotta

perrotta commented:
Thank you Slava!
@sarafiorendi


Slava Krutelyov <notifications@github.com> ha scritto:

> slava77 commented on this pull request.
>
>
>
>> @@ -194,6 +195,7 @@ class HcalNoiseSummary
>    double rechitEnergy_;
>    double rechitEnergy15_;
>    double calibCharge_;
> +  double lasmonCharge_;
>
> this variable is not initialized.
> Since I've added the MET filter path monitoring a few days ago, now  
> we find that this leads to random differences when reading AOD  
> produced by a release before 100X.
> E.g. https://github.com/cms-sw/cmssw/pull/25320#issuecomment-441199764
>
> @perrotta
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/21152#pullrequestreview-177898616
